game/src/engine/audio/AudioService.cpp

```diff
@@ -1,38 +1,200 @@
 #include "AudioService.h"
 
+#include <AL/al.h>
 #include <AL/alc.h>
-#include <AudioFile.h>
 
+#include <iostream>
+#include <iterator>
+#include <memory>
 #include <object_ptr.hpp>
+#include <string>
+#include <utility>
 
+#include "AudioFile.h"
 #include "engine/core/debug/Log.h"
 #include "engine/input/InputService.h"
 #include "engine/service/Service.h"
 #include "engine/service/ServiceProvider.h"
+#include "stb/stb_vorbis.c"
 
 // the system's default output device will be used
-const ALCchar* default_device = nullptr;
-ALCint context_attributes = 0;
+const ALCchar* kDefaultDevice = nullptr;
+ALCint kContextAttributes = 0;
 
-AudioService::AudioService()
-    : audio_device_(alcOpenDevice(default_device)),
-      audio_context_(alcCreateContext(audio_device_, 0))
+// audio directories
+// note: since diagetic sound (3d sound) can only be in mono
+// we'll want all sfx to be mono too (music is fine in stereo)
+const std::string kSfxDirectory = "resources/audio/sfx/";
+const std::string kMusicDirectory = "resources/audio/music/";
+
+// for debugging
+const std::string kTestFileName = "professional_test_audio.ogg";
+
+void AudioService::PlayOneShot(std::string file_name, float gain)
 {
+    // check if source already exists;
+    // don't want to keep adding buffers for the same sound
+    if (active_sources_.count(file_name) == 0)
+        AddSource(file_name);
+
+    ALuint source;
+    source = active_sources_[file_name].first;
+
+    // set gain if provided (defaults to 1.f)
+    alSourcef(source, AL_GAIN, gain);
+
+    Log::debug("Playing audio: {}", file_name);
+
+    alSourcePlay(source);
 }
 
-void AudioService::OnInit()
+void AudioService::PlayLoop(std::string file_name, float gain)
+{
+    // TODO
+}
+
+void AudioService::StopPlayback(std::string file_name)
+{
+    ALuint source = active_sources_[file_name].first;
+    alSourceStop(source);
+}
+
+void AudioService::StopAllPlayback()
 {
-    // check audio device
-    if (!audio_device_)
+    for (auto& pair : active_sources_)
     {
-        Log::warning("[AudioService] couldn't open audio device.");
+        alSourceStop(pair.second.first);
     }
+}
+
+/* ----- getters / setters / helpers ----- */
+
+AudioFile AudioService::LoadAudioFile(std::string file_name, bool is_looping)
```

Comment:
nit: perhaps rename "is_looping" to "is_mono" or something related to whether it's 3d audio?

Reply:
yeah, i was thinking of just making a separate function to distinguish sound effects from music at some point since there are sound effects that we would want to loop and be 3d (the sound of cars actually driving, tire screeching, etc)

Reply:
yeah that makes sense!